src/frontend/src/optimizer/plan_node/logical_source.rs

```diff
@@ -356,7 +357,10 @@ impl ToBatch for LogicalSource {
 impl ToStream for LogicalSource {
     fn to_stream(&self, _ctx: &mut ToStreamContext) -> Result<PlanRef> {
         let mut plan: PlanRef = StreamSource::new(self.clone()).into();
-        if let Some(row_id_index) = self.core.row_id_index && self.core.gen_row_id{
+        if let Some(catalog) = self.source_catalog() && !catalog.watermark_descs.is_empty(){
+            plan = StreamWatermarkFilter::new(plan, catalog.watermark_descs.clone()).into();
```

Comment:
How to test this in e2e or planner test? 🤔

(Perhaps after supporting watermark on append-only tables?)

Reply:
Added a planner test.

Reply:
we can check the state cleaning in the e2e test because our state cleaning is based on the consistent `range_delete`